### PR TITLE
Compile wasmtime with effect-handlers/wasm-tools:func-ref-2

### PR DESCRIPTION
**Pull request by dhil**
This patch makes `wasmtime` compile with branch `func-ref-2` in our `wasm-tools` fork. There are many todos left to be completed.

The test suite compiles and passes.

### CODE REVIEW

File: cranelift/wasm/src/translation_utils.rs [hunk 1]
```diff
@@ -123,12 +144,18 @@ pub fn block_with_params<PE: TargetEnvironment + ?Sized>(
             wasmparser::ValType::F64 => {
                 builder.append_block_param(block, ir::types::F64);
             }
-            wasmparser::ValType::ExternRef | wasmparser::ValType::FuncRef => {
-                builder.append_block_param(block, environ.reference_type(ty.try_into()?));
+            wasmparser::ValType::Ref(rt) => {
+                match rt.heap_type {
+                    wasmparser::HeapType::Func | wasmparser::HeapType::Extern => {
+                        builder.append_block_param(block, environ.reference_type(rt.heap_type.try_into()?));
```
Comment (CosineP) [inline cranelift/wasm/src/translation_utils.rs:150]: surely the reference type constructor will need to know nullability as well?

File: crates/cranelift/src/func_environ.rs [hunk 2]
```diff
@@ -1317,11 +1317,12 @@ impl<'module_environment> cranelift_wasm::FuncEnvironment for FuncEnvironment<'m
         mut pos: cranelift_codegen::cursor::FuncCursor<'_>,
         index: cranelift_wasm::GlobalIndex,
     ) -> WasmResult<ir::Value> {
-        debug_assert_eq!(
-            self.module.globals[index].wasm_ty,
-            WasmType::ExternRef,
-            "We only use GlobalVariable::Custom for externref"
-        );
+        // TODO(dhil) fixme: need to check that wasm_ty is indeed a Ref.
```
Comment (CosineP) [inline crates/cranelift/src/func_environ.rs]: I don't know anything about GlobalVariable::Custom, but if the message is true we can probably leave this as is but with EXTERN_REF

File: crates/cranelift/src/lib.rs [hunk 3]
```diff
@@ -206,10 +207,10 @@ fn func_signature(
...
-    match wasm_ty {
-        cranelift_wasm::WasmType::FuncRef => pointer_type,
-        cranelift_wasm::WasmType::ExternRef => match pointer_type {
+fn reference_type(wasm_ht: cranelift_wasm::WasmHeapType, pointer_type: ir::Type) -> ir::Type {
+    match wasm_ht {
+        cranelift_wasm::WasmHeapType::Func => pointer_type,
+        cranelift_wasm::WasmHeapType::Extern => match pointer_type {
             ir::types::I32 => ir::types::R32,
             ir::types::I64 => ir::types::R64,
             _ => panic!("unsupported pointer type"),
```
Comment (CosineP) [inline crates/cranelift/src/lib.rs:216]: presumably need a case for Index

File: crates/environ/src/module.rs [hunk 4]
```diff
@@ -428,7 +435,7 @@ impl ModuleTranslation<'_> {
 
             // If this is not a funcref table, then we can't support a
             // pre-computed table of function indices.
-            if self.module.table_plans[segment.table_index].table.wasm_ty != WasmType::FuncRef {
+            if !is_func_ref(self.module.table_plans[segment.table_index].table.wasm_ty) {
```
Comment (CosineP) [inline crates/environ/src/module.rs]: You can use == here just fine, just compare to WasmHeapType::Func

File: crates/runtime/src/table.rs [hunk 5]
```diff
@@ -163,11 +163,11 @@ pub enum Table {
...
-fn wasm_to_table_type(ty: WasmType) -> Result<TableElementType> {
-    match ty {
-        WasmType::FuncRef => Ok(TableElementType::Func),
-        WasmType::ExternRef => Ok(TableElementType::Extern),
-        ty => bail!("invalid table element type {:?}", ty),
+fn wasm_to_table_type(rt: WasmRefType) -> Result<TableElementType> {
+    match rt.heap_type {
+        WasmHeapType::Func => Ok(TableElementType::Func),
+        WasmHeapType::Extern => Ok(TableElementType::Extern),
+        ht => bail!("invalid table element type {:?}", ht),
```
Comment (CosineP) [inline crates/runtime/src/table.rs:171]: Need (a todo for) Index

File: crates/types/src/lib.rs [hunk 6]
```diff
@@ -87,10 +161,10 @@ pub struct WasmFuncType {
 impl WasmFuncType {
     #[inline]
     pub fn new(params: Box<[WasmType]>, returns: Box<[WasmType]>) -> Self {
-        let externref_params_count = params.iter().filter(|p| **p == WasmType::ExternRef).count();
+        let externref_params_count = params.iter().filter(|p| match **p { WasmType::Ref(WasmRefType { heap_type: WasmHeapType::Extern, .. }) => true, _ => false }).count();
```
Comment (CosineP) [inline crates/types/src/lib.rs]: `p.heap_type == WasmHeapType::Extern` - even more succinct than the old version

File: crates/types/src/lib.rs [hunk 7]
```diff
@@ -87,10 +161,10 @@ pub struct WasmFuncType {
...
         let externref_returns_count = returns
             .iter()
-            .filter(|r| **r == WasmType::ExternRef)
+            .filter(|r| match **r { WasmType::Ref(WasmRefType { heap_type: WasmHeapType::Extern, .. }) => true, _ => false })
```
Comment (CosineP) [inline crates/types/src/lib.rs]: Speaking of which do non-nullable extern refs exist?

File: crates/wasmtime/src/externals.rs [hunk 8]
```diff
@@ -459,7 +465,7 @@ impl Table {
 
     fn _new(store: &mut StoreOpaque, ty: TableType, init: Val) -> Result<Table> {
         let wasmtime_export = generate_table_export(store, &ty)?;
-        let init = init.into_table_element(store, ty.element())?;
+        let init = init.into_table_element(store, ValType::Ref(ty.element()))?;
```
Comment (CosineP) [inline crates/wasmtime/src/externals.rs]: maybe into_table_element should accept a RefType?

File: crates/wasmtime/src/func/typed.rs [hunk 9]
```diff
@@ -321,7 +321,7 @@ unsafe impl WasmTy for Option<ExternRef> {
```
Comment (CosineP) [inline crates/wasmtime/src/func/typed.rs:321]: Where does ExternRef come from here?

File: crates/wasmtime/src/types.rs [hunk 10]
```diff
@@ -47,8 +47,10 @@ impl fmt::Display for ValType {
             ValType::F32 => write!(f, "f32"),
             ValType::F64 => write!(f, "f64"),
             ValType::V128 => write!(f, "v128"),
-            ValType::ExternRef => write!(f, "externref"),
-            ValType::FuncRef => write!(f, "funcref"),
+            ValType::Ref(rt) => write!(f, "ref {}", rt),
```
Comment (CosineP) [inline crates/wasmtime/src/types.rs]: "ref " should presumably be handled by Ref printing (and parens)

File: crates/wasmtime/src/types.rs [hunk 11]
```diff
@@ -90,14 +92,88 @@ impl ValType {
...
+            WasmType::Bot => Self::Bot,
         }
     }
 }
 
-// External Types
+// Reference types
+/// A list of all possible value types in WebAssembly.
+#[derive(Debug, Clone, Hash, Eq, PartialEq)]
+pub struct RefType {
```
Comment (CosineP) [inline crates/wasmtime/src/types.rs:102]: Just curious, why does this get defined twice?

**Review by CosineP — APPROVED** [review]
This is great, and looks like the tests pass too?   Cool.  Alright looks like we have our work cut out for us :)

File: crates/types/src/lib.rs (hunk 7 above)
Comment (dhil) [reply crates/types/src/lib.rs]: I don't think so, as `externref` and `funcref` are interpreted as nullable refs c.f. https://github.com/WebAssembly/function-references/blob/main/proposals/function-references/Overview.md#Reference-Types.

File: crates/wasmtime/src/func/typed.rs (hunk 9 above)
Comment (dhil) [reply crates/wasmtime/src/func/typed.rs:321]: It is defined in https://github.com/bytecodealliance/wasmtime/blob/main/crates/wasmtime/src/ref.rs#L10

**Pull request merged**